File: 03 Leet - Hard/42-trapping-rain-water.js
/*
Given n non-negative integers representing an elevation map where the width of each bar is 1, compute how much water it is able to trap after raining.
*/

/**
 * @param {number[]} heights
 * @return {number}
 */

var peakIndex = 0;
// var heightsArray = [2, 0, 2];
// var heightsArray = [4, 2, 3];
var heightsArray = [0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1];

var trap = function(heights) {
  let total = 0;
//   let stupidTest = [2, 0, 2];

  if (heights[1] < heights[0]) {
    total += findWaterBetweenPeaks(peakIndex, heights);
  } else {
    peakIndex = 1;
  }

  for (i = peakIndex; i < heights.length; i++) {
    if (heights[i] > heights[i + 1] && heights[i] > heights[i - 1]) {
      // this is a peak
      peakIndex = i;
      total += findWaterBetweenPeaks(peakIndex, heights);
    }
  }

  return total;
};

var findWaterBetweenPeaks = function(firstPeakIndex, heights) {
  let totalBetweenPeaks = 0;
  let waterLevels = new Array(heights[firstPeakIndex]).fill(0);

  let i = firstPeakIndex + 1;
  //While in a valley
  while (i < heights.length) {
    if (i === heights.length - 1 && heights[i] > heights[i - 1]) {
      // this is a peak
      break;
    }
    if (
      (heights[i] > heights[i + 1] && heights[i] > heights[i - 1]) ||
      i === heights.length - 1
    ) {
      // this is a peak
      break;
    }

    for (j = heights[firstPeakIndex] - 1; j > heights[i] - 1; j--) {
      waterLevels[j] += 1;
    }

    i++;
  }

  for (k = heights[i]; k < waterLevels.length; k++) {
    waterLevels[k] = 0;
  }

//   console.log('waterLevel', waterLevels);
  totalBetweenPeaks = waterLevels.reduce((a, b) => a + b, 0);

//   console.log('totalBetweenPeaks', totalBetweenPeaks);

  peakIndex = i;
  return totalBetweenPeaks;
};

console.log(trap(heightsArray)); // 5

File: NeetCode +75/02 - Two Pointers/42-trapping-rain-water.py
'''
Given n non-negative integers representing an elevation map where the width of each bar is 1, compute how much water it can trap after raining.
'''

from time import time
from typing import List


class Solution:
    def trap(self, height: List[int]) -> int:
        res = 0
        leftMax, rightMax = 0, 0
        l, r = 0, len(height) - 1
        while l < r:
            # Water level is limited by shorter wall
            if height[l] < height[r]:
                if height[l] > leftMax:
                    leftMax = height[l]
                else:
                    res += leftMax - height[l]
                l += 1
            else:
                if height[r] > rightMax:
                    rightMax = height[r]
                else:
                    res += rightMax - height[r]
                r -= 1

        return res

    def reference(self, height: List[int]) -> int:
        if not height:
            return 0

        l, r = 0, len(height) - 1
        leftMax, rightMax = height[l], height[r]
        res = 0
        while l < r:
            if leftMax < rightMax:
                l += 1
                leftMax = max(leftMax, height[l])
                res += leftMax - height[l]
            else:
                r -= 1
                rightMax = max(rightMax, height[r])
                res += rightMax - height[r]
        return res

    def quantify(self, test_cases, runs=50000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.trap(case))
                else:
                    self.trap(case)
        print(f'Runtime for our solution: {time() - sol_start}\n')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(case))
                else:
                    self.reference(case)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [[0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1], [4, 2, 0, 3, 2, 5]]
    test.quantify(test_cases)
